KerbalStuff/blueprints/api.py

```diff
@@ -463,6 +464,40 @@ def change_password(username: str) -> Union[Dict[str, Any], Tuple[Union[str, Any
     return {'error': True, 'reason': pw_message}
 
 
+@api.route("/api/user/<username>/delete", methods=['POST'])
+@with_session
+@user_required
+@json_output
+def delete(username: str) -> Dict[str, Any]:
+    deletable = False
+    if current_user:
+        if current_user.admin:
+            deletable = True
+        if current_user.username == username:
+            deletable = True
+    if not deletable:
+        return {'error': True, 'reason': 'Unauthorized'}
+
+    form_username = request.form.get('username')
+    if form_username != username:
+        return {'error': True, 'reason': 'Wrong username'}
```

Comment:
Hmm, it doesn't feel right to assume form inputs in an `/api/` route. An "API" should be usable by whatever random JSON clients people want to write (e.g. Netkan or Nertea's publishing scripts), not just the front-end.

Instead of checking whether there's a match after the user clicks the Delete confirmation button, could we try this?

- The Delete button in the popup starts out disabled (but Cancel is always enabled)
- An event handler in the username box checks whether its value is correct, and then enables or disables the Delete button based on that

That way the user could only click the button after they type the name. If some admin wants to delete users with an API script independently, the confirmation flow would not be relevant to them anyway.

Reply:
> Hmm, it doesn't feel right to assume form inputs in an /api/ route. An "API" should be usable by whatever random JSON clients people want to write (e.g. Netkan or Nertea's publishing scripts), not just the front-end.

This is pretty normal and standard, we do this in a lot of API routes. That's why they are POSTs. You need some way to transmit data to the server.
I'm not aware of any HTTP client that is incapable of including form data in a POST request. Netkan only reads from the API and doesn't write anything, thus it also doesn't use POSTs. But I'm very confident that WebClient could include form data if it needed to.

Here it is primarily meant as a "confirmation" feature so you don't delete your account accidentally. Whether that's useful for APIs is debatable I guess, I'm okay with removing it. I'm pretty sure that just came from adopting the password change mechanism.

Reply:
Oh right, the `api.md` documentation does say:

> Submit all POSTS with the request body encoded as
> [multipart/form-data](https://www.ietf.org/rfc/rfc2388.txt). Your HTTP library
> of choice probably handles that for you. All responses are JSON.

... so using a form input is fine. I guess my objection is just to this one specific form input, then.

Reply:
Added an implementation of this suggestion in #417, with script code enabling/disabling the button.